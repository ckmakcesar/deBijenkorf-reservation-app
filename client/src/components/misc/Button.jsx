import React, { useState } from 'react';
import PropTypes from 'prop-types';
import clsx from 'clsx';

import styles from '../../styles/misc/Button.module.css';

const Button = ({
  onClick,
  text,
  disabled,
  isSubmit,
  formId,
  boxIconClassName,
  tone,
  id,
}) => {
  const [hover, setHover] = useState(false);

  const handleClick = (e) => {
    if (!disabled) {
      onClick(e);
    }
  };
  const handleMouseEnter = () => {
    if (!disabled) {
      setHover(true);
    }
  };
  const handleMouseLeave = () => {
    if (!disabled) {
      setHover(false);
    }
  };

  const conditionalColorTone = (disabled
    ? styles.disabled
    : {
      [styles.hover]: hover && !tone,
      [styles[tone]]: !hover && tone,
      [styles[tone + 'Hover']]: hover && tone,
    }
  );

  return (
    <div
      className={clsx(
        styles.buttonContainer,
        conditionalColorTone,
        disabled && styles.disabledBorder
      )}
      onClick={handleClick}
      onMouseEnter={handleMouseEnter}
      onMouseLeave={handleMouseLeave}
    >
      <span
        className={clsx(
          styles.buttonOuter,
          !tone && hover && styles.buttonOuterHover,
          conditionalColorTone
        )}
      >
        {boxIconClassName
          ? <i
            className={clsx(
              'bx', styles.buttonIcon, boxIconClassName,
              conditionalColorTone
            )}
          />
          : null
        }
        <input
          {...(id ? { id: `Button-${id}` } : null)} // optional id - should be unique
          className={clsx(
            styles.button,
            boxIconClassName && styles.buttonReducedLeftPadding,
            conditionalColorTone
          )}
          type={isSubmit ? 'submit' : 'button'}
          form={isSubmit && formId ? formId : ''}
          value={text.toUpperCase()}
        />
      </span>
    </div >
  );
};

// *********************************************************************************************************************
// PROPS
// *********************************************************************************************************************
Button.propTypes = {
  onClick: PropTypes.func,
  text: PropTypes.string,
  disabled: PropTypes.bool,
  isSubmit: PropTypes.bool,
  formId: PropTypes.string,
  boxIconClassName: PropTypes.string,
  tone: PropTypes.string,
  id: PropTypes.string,
};

Button.defaultProps = {
  onClick: () => { },
  text: 'OK',
  disabled: false,
  isSubmit: false,
  formId: '',
  boxIconClassName: '',
  tone: '',
  id: '',
};

export default Button;
