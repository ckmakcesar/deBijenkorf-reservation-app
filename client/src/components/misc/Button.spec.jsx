import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import Button from './Button';

describe('component:Button', () => {
  it('Should render Button with every optional props absent', () => {
    const wrapper = shallow(
      <Button />,
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('Should render Button with every optional props present', () => {
    const wrapper = shallow(
      <Button
        onClick={jest.fn()}
        text='Click to Submit'
        disabled={false}
        isSubmit={true}
        formId='form001'
        boxIconClassName='bx-send'
        tone='secondary'
      />,
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
