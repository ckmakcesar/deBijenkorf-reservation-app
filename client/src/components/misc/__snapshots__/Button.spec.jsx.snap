// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component:Button Should render Button with every optional props absent 1`] = `
<div
  className="buttonContainer"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <span
    className="buttonOuter"
  >
    <input
      className="button buttonReducedLeftPadding"
      form=""
      type="button"
      value="OK"
    />
  </span>
</div>
`;

exports[`component:Button Should render Button with every optional props present 1`] = `
<div
  className="buttonContainer secondary"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <span
    className="buttonOuter secondary"
  >
    <i
      className="bx buttonIcon bx-send secondary"
    />
    <input
      className="button buttonReducedLeftPadding secondary"
      form="form001"
      type="submit"
      value="CLICK TO SUBMIT"
    />
  </span>
</div>
`;
